spec/20-AMQP_FORMAT.md

```diff
@@ -1,3 +1,132 @@
+business messaging. It defines a binary wire-level protocol that allows for the
+reliable exchange of business messages between two parties. AMQP can be used as
+a protocol for asynchronous communication between components of an application.
+From the distributed tracing and telemetry correlation perspective it is a
+known problem to be able to correlate a component that placed message and
+component that processed it later. This specification describes how trace
+context MUST be encoded into AMQP messages.
+
+## Trace context fields placement in a message
+
+AMQP defines message as a payload with the additional annotations sections.
+There are two annotation sections this specification refers -
+"application-properties" and "message-annotations". See
+[3.2](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format)
+of *OASIS Advanced Message Queuing Protocol (AMQP) Version 1.0, Part 3:
+Messaging* (AMQP specification).
+
+AMQP message section "application-properties" is immutable collection of
+properties that is defined by message publisher and can be read by the message
+consumer. Message brokers cannot mutate those properties. Section
+"message-annotations" is designed for message brokers to use and can be mutated
+during the message processing.
+
+Fields `traceparent` and `tracestate` SHOULD be added to the message in the
+application-properties section by message publisher. Once the message has been
+created, it’s no longer permissible to edit the bare message. So if it were
+necessary to annotate the message inside the middleware as it transits, that
+MUST happen in the “message-annotations” section, using the
+“application-properties” as a base. Message reader SHOULD construct the full
+trace context by reading `traceparent` and `tracestate` fields from the
+“message-annotations” first and if not exist - from “application-properties”.
+
+## Trace context and failed read/write operations
+
+This specification defines how to propagate context from publisher thru broker
+to ultimate reader. It does not define any additional transport level correlation
+constructs that can be used to investigate failed publish or read operations.
+
+It is recommended, however, for AMQP implementations to make the best effort
+attempt to read the trace context from the message and use it while reporting
+such problems.
+
+TODO: do we need to define a way to specify trace context of read operation? Is
+there anything in AMQP protocol that can be used to carry this context?
+
+## `traceparent` AMQP format
+
+The field `traceparent` MUST be encoded and decoded using [binary
+protocol](..\extension-binary.html) and stored as a binary type defined in
+section
+[1.6.19](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-binary)
+of AMQP specification.
+
+Property name MUST be `traceparent` - all lowercase without delimiters.
```

Comment:
from: @calohmn

Considering the case that the `traceparent` and `tracestate` fields are transported via _message-annotations_:

As for the base type of the message-annotations, the AMQP spec says (http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-annotations):

> The annotations type is a map where the keys are restricted to be of type symbol or of type ulong. All ulong keys, and all symbolic keys except those beginning with "x-" are reserved. Keys beginning with "x-opt-" MUST be ignored if not understood. On receiving an annotation key which is not understood, and which does not begin with "x-opt", the receiving AMQP container MUST detach the link with a not-implemented error.

Concerning the names of `message-annotations`:

> A registry of defined annotations and their meanings is maintained [[AMQPMESSANN](http://www.amqp.org/specification/1.0/message-annotations)].

The question here: Are the names of the `traceparent` and `tracestate` fields supposed to be somehow registered as reserved with OASIS. Otherwise it looks to me as if the prefix "x-opt-" should be used.

Reply:
Also created this issue: https://github.com/w3c/trace-context-amqp/issues/2

Reply:
As I wrote in the issue, we can register this fields in AMQP TC